crates/client/src/socks.rs

```diff
@@ -3,59 +3,102 @@ use anyhow::{Result, anyhow};
 use http_client::Url;
 use tokio_socks::tcp::{Socks4Stream, Socks5Stream};
 
+/// Identification to a Socks V4 Proxy
+struct Socks4Identification<'a> {
+    user_id: &'a str,
+}
+
+/// Authorization to a Socks V5 Proxy
+struct Socks5Authorization<'a> {
+    username: &'a str,
+    password: &'a str,
+}
+
+/// Socks Proxy Protocol Version
+///
+/// V4 allows idenfication using a user_id
+/// V5 allows authorization using a username and password
+enum SocksVersion<'a> {
+    V4(Option<Socks4Identification<'a>>),
+    V5(Option<Socks5Authorization<'a>>),
+}
+
 pub(crate) async fn connect_socks_proxy_stream(
     proxy: Option<&Url>,
     rpc_host: (&str, u16),
 ) -> Result<Box<dyn AsyncReadWrite>> {
-    let stream = match parse_socks_proxy(proxy) {
-        Some((socks_proxy, SocksVersion::V4)) => {
-            let stream = Socks4Stream::connect_with_socket(
-                tokio::net::TcpStream::connect(socks_proxy).await?,
-                rpc_host,
-            )
-            .await
-            .map_err(|err| anyhow!("error connecting to socks {}", err))?;
-            Box::new(stream) as Box<dyn AsyncReadWrite>
+    // Connect directly when there is no proxy
+    let Some(proxy) = proxy else {
+        return Ok(Box::new(tokio::net::TcpStream::connect(rpc_host).await?));
+    };
+
+    let Some((socks_proxy, version)) = parse_socks_proxy(proxy) else {
```

Comment:
We can remove this nesting with

```rs
use anyhow::Context as _;

let (socks_proxy, version) = parse_socks_proxy(proxy).with_context(|| format!("Parsing proxy url '{proxy}'"))?;
// or use `.context("parsing proxy url")?` if `url` is not secure to print due to the risk of leaking the credentials.
```

and might be good to do `.context` on more `?` around, such as `TcpStream::connect` or maybe even `.map_err(|err| anyhow!`?

Reply:
Mostly a stylistic choice, but I prefer the explicitness of the return when parsing the proxy url, since it's quite critical that we don't use a fallback. However if context adheres better to Zed coding styles I'll gladly change it!

I'll add a `.map_err` to the `TcpStream::connect`, not sure why I didn't in the first place :)

Reply:
I would prefer `.context`/`.with_context` over `.map_err` and `?` usage anyway as indeed, mostly the Zed "code style" and the type system being rather explicit itself already.

Reply:
I left this block intact for now, would you like me to replace it with a `map_err` ?

```rust
let Some((socks_proxy, version)) = parse_socks_proxy(proxy) else {
    // If parsing the proxy URL fails, we must avoid falling back to an insecure connection.
    // SOCKS proxies are often used in contexts where security and privacy are critical,
    // so any fallback could expose users to significant risks.
    return Err(anyhow!("Parsing proxy url failed"));
};
```

Reply:
I think it's fine, thank you for the rest of the fixes.